File: sample/src/main/java/vn/tiki/sample/login/LoginPresenter.java
package vn.tiki.sample.login;

import android.support.annotation.NonNull;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;
import vn.tiki.architecture.mvp.ViewAction;
import vn.tiki.sample.model.UserModel;
import vn.tiki.sample.mvp.rx.RxBasePresenter;
import vn.tiki.sample.util.EmailValidator;
import vn.tiki.sample.util.PasswordValidator;

class LoginPresenter extends RxBasePresenter<LoginView> {

  @NonNull private final PublishSubject<String> emailInputs = PublishSubject.create();
  @NonNull private final PublishSubject<String> passwordInputs = PublishSubject.create();
  @NonNull private final PublishSubject<Object> loginClicks = PublishSubject.create();
  @NonNull private final PublishSubject<Boolean> networkStatusChanges = PublishSubject.create();

  private String email;
  private String password;

  LoginPresenter(@NonNull final UserModel userModel) {
    final EmailValidator emailValidator = new EmailValidator();
    final PasswordValidator passwordValidator = new PasswordValidator();

    final Observable<Boolean> emailValidations = emailInputs
        .doOnNext(s -> email = s)
        .map(emailValidator::validate);

    final Observable<Boolean> passwordValidations = passwordInputs
        .doOnNext(s -> password = s)
        .map(passwordValidator::validate);

    disposeOnDestroy(emailValidations
        .distinctUntilChanged()
        .subscribe(valid -> {
          if (valid) {
            getViewOrThrow().hideValidationEmailError();
          } else {
            getViewOrThrow().showValidationEmailError();
          }
        }));

    disposeOnDestroy(passwordValidations
        .distinctUntilChanged()
        .subscribe(valid -> {
          if (valid) {
            getViewOrThrow().hideValidationPasswordError();
          } else {
            getViewOrThrow().showValidationPasswordError();
          }
        }));

    disposeOnDestroy(networkStatusChanges.distinctUntilChanged()
        .subscribe(isConnected -> {
          if (isConnected) {
            getViewOrThrow().hideNetworkError();
          } else {
            getViewOrThrow().showNetworkError();
          }
        }));

    disposeOnDestroy(Observable.combineLatest(
        emailValidations,
        passwordValidations,
        networkStatusChanges,
        (validEmail, validPassword, isConnected) -> validEmail && validPassword && isConnected)
        .subscribe(validSubmit -> {
          if (validSubmit) {
            getViewOrThrow().enableSubmit();
          } else {
            getViewOrThrow().disableSubmit();
          }
        }));

    final Observable<LoginResult> authentications =
        loginClicks.switchMap(__ ->
            userModel.login(email, password)
                .map(aBoolean -> {
                  if (aBoolean) {
                    return LoginResult.success();
                  } else {
                    throw new Exception("authentication failed");
                  }
                })
                .onErrorReturn(LoginResult::error)
                .subscribeOn(Schedulers.io())
                .startWith(LoginResult.inFlight()));

    disposeOnDestroy(authentications
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(loginResult -> {
          final ViewAction<LoginView> viewAction;
          if (loginResult.isInFlight()) {
            viewAction = view -> {
              view.disableSubmit();
              view.hideAuthenticationError();
              view.hideNetworkError();
              view.showLoading();
            };
          } else if (loginResult.isError()) {
            viewAction = view -> {
              view.hideLoading();
              view.showAuthenticationError();
              view.enableSubmit();
            };
          } else {
            viewAction = LoginView::showLoginSuccess;
          }
          sendToView(viewAction);
        }));
  }

  void onNetworkStatusChanged(boolean connected) {
    networkStatusChanges.onNext(connected);
  }

  void onInputEmail(String email) {
    emailInputs.onNext(email);
  }

  void onInputPassword(String password) {
    passwordInputs.onNext(password);
  }

  void onClickLogin() {
    loginClicks.onNext(new Object());
  }
}

File: sample/src/main/java/vn/tiki/sample/login/LoginPresenter.kt
package vn.tiki.sample.login

import android.text.Editable
import butterknife.OnClick
import butterknife.OnTextChanged
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Function3
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import vn.tiki.sample.R
import vn.tiki.sample.di.ActivityScope
import vn.tiki.sample.model.UserModel
import vn.tiki.sample.mvp.rx.RxBasePresenter
import vn.tiki.sample.util.EmailValidator
import vn.tiki.sample.util.PasswordValidator
import vn.tiki.sample.util.subscribe
import javax.inject.Inject

@ActivityScope
class LoginPresenter @Inject constructor(userModel: UserModel) : RxBasePresenter<LoginView>() {

  private val emailInputs = PublishSubject.create<String>()
  private val passwordInputs = PublishSubject.create<String>()
  private val loginClicks = PublishSubject.create<Any>()
  private val networkStatusChanges = PublishSubject.create<Boolean>()

  private var email: String? = null
  private var password: String? = null

  init {
    val emailValidator = EmailValidator()
    val passwordValidator = PasswordValidator()

    val emailValidations = emailInputs
        .doOnNext { s -> email = s }
        .map(emailValidator::validate)

    val passwordValidations = passwordInputs
        .doOnNext { s -> password = s }
        .map(passwordValidator::validate)

    emailValidations
        .distinctUntilChanged()
        .subscribe(this) { valid ->
          if (valid!!) {
            viewOrThrow.hideValidationEmailError()
          } else {
            viewOrThrow.showValidationEmailError()
          }
        }

    passwordValidations
        .distinctUntilChanged()
        .subscribe(this) {
          if (it) {
            viewOrThrow.hideValidationPasswordError()
          } else {
            viewOrThrow.showValidationPasswordError()
          }
        }

    Observable.combineLatest<Boolean, Boolean, Boolean, Boolean>(emailValidations,
        passwordValidations,
        networkStatusChanges,
        Function3<Boolean, Boolean, Boolean, Boolean> { validEmail, validPassword, isConnected -> validEmail && validPassword && isConnected })
        .subscribe { validSubmit ->
          if (validSubmit) {
            view?.enableSubmit()
          } else {
            view?.disableSubmit()
          }
        }

    val authentications = loginClicks.switchMap { _ ->
      userModel.login(email!!, password!!)
          .map { LoginSuccess as LoginState }
          .onErrorReturn { LoginError(it) }
          .subscribeOn(Schedulers.io())
          .startWith(LoginInFlight)
    }

    authentications
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(this) { result ->
          when (result) {
            is LoginInFlight -> sendToView { view ->
              view.disableSubmit()
              view.hideAuthenticationError()
              view.showLoading()
            }
            is LoginError -> sendToView { view ->
              view.hideLoading()
              view.showAuthenticationError()
              view.enableSubmit()
            }
            else -> sendToView(LoginView::showLoginSuccess)
          }
        }
  }

  fun onNetworkStatusChanged(connected: Boolean) {
    networkStatusChanges.onNext(connected)
  }

  @OnTextChanged(value = R.id.etEmail, callback = OnTextChanged.Callback.AFTER_TEXT_CHANGED)
  fun onInputEmail(editable: Editable) {
    emailInputs.onNext(editable.toString())
  }

  @OnTextChanged(value = R.id.etPassword, callback = OnTextChanged.Callback.AFTER_TEXT_CHANGED)
  fun onInputPassword(editable: Editable) {
    passwordInputs.onNext(editable.toString())
  }

  @OnClick(R.id.btLogin)
  fun onClickLogin() {
    loginClicks.onNext(Any())
  }
}
